### PR TITLE
Consistent process test procedure paths

### PR DESCRIPTION
**Pull request by sebeder**
[Trello Card](https://trello.com/c/tsMEvsdY/585-testgenerierung-konsistente-ber%C3%BCcksichtigung-von-bedingungen)

### CODE REVIEW

File: bundles/specmate-testspecification/src/com/specmate/testspecification/internal/generators/ProcessTestCaseGenerator.java [hunk 1]
```diff
@@ -86,11 +91,41 @@ protected void generateParameters() {
...
 	protected void generateTestCases() throws SpecmateException {
-		List<GraphPath<IModelNode, ProcessConnection>> allPaths = getAllPaths();
-		List<GraphPath<IModelNode, ProcessConnection>> filteredPaths = filterDuplicatePaths(allPaths);
+		Set<IModelNode> startNodes = getStartNodes();
+		Set<IModelNode> endNodes = getEndNodes();
+		
+		List<GraphPath<IModelNode, ProcessConnection>> paths = getAllPathsExact(startNodes, endNodes);
+
+		Set<ProcessConnection> coveredConnections = paths.stream()
+				.flatMap(g -> g.getEdgeList().stream())
```
Comment (junkerm) [inline bundles/specmate-testspecification/src/com/specmate/testspecification/internal/generators/ProcessTestCaseGenerator.java]: rename g to path to make clear you are iterating over paths

File: bundles/specmate-testspecification/src/com/specmate/testspecification/internal/generators/ProcessTestCaseGenerator.java [hunk 2]
```diff
@@ -104,66 +139,211 @@ private TestParameter createAndAddTestParameterIfNecessary(String testParameterN
...
+		Set<String> conflictingConditions = decisionNodes.stream()
+				.flatMap(n -> n.getOutgoingConnections().stream())
+				.filter(c -> !hasCondition(c, condition))
+				.map(c -> (ProcessConnection) c)
+				.map(c -> c.getCondition())
+				.collect(Collectors.toSet());
+		return conflictingConditions;
+	}
+	
+	private static boolean hasCondition(IModelNode node, String condition) {
```
Comment (junkerm) [inline bundles/specmate-testspecification/src/com/specmate/testspecification/internal/generators/ProcessTestCaseGenerator.java:171]: There is also a non-static hasCondition method, I find this a bit confusing to have static and non-static methods with the same name in different places in the file

File: bundles/specmate-testspecification/src/com/specmate/testspecification/internal/generators/ProcessTestCaseGenerator.java [hunk 3]
```diff
@@ -104,66 +139,211 @@ private TestParameter createAndAddTestParameterIfNecessary(String testParameterN
...
+		return false;
+	}
+	
+	/**
+	 * Gets the current graph without edges containing conflicting conditions to the given condition.
+	 * 
+	 * @param condition the condition
+	 * @return the graph without conflicting conditions to the given condition
+	 */
+	private DirectedGraph<IModelNode, ProcessConnection> getGraph(String condition) {
```
Comment (junkerm) [inline bundles/specmate-testspecification/src/com/specmate/testspecification/internal/generators/ProcessTestCaseGenerator.java]: could you pick a method name that is more descriptive of what is going on?

**Review by junkerm — CHANGES_REQUESTED** [review]
It looks good. I did two test models and got the expected results.
grafik
grafik

Could you adapt the unit tests so that they run and capture the additional contstraints?

**Comment by sebeder** [comment]
Hi @junkerm ! I adressed your comments. Would you review it again? Thank!

**Review by junkerm — APPROVED** [review]
tests look good!

**Pull request merged**